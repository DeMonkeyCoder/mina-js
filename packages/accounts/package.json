{
	"name": "@mina-js/accounts",
	"type": "module",
	"version": "0.1.1",
	"module": "dist/index.js",
	"types": "dist/index.d.ts",
	"exports": {
		".": {
			"types": "./dist/index.d.ts",
			"default": "./dist/index.js"
		}
	},
	"files": ["dist"],
	"scripts": {
		"build": "tsc --noEmit && tsup",
		"test": "bun test",
		"cleanup": "rimraf dist .turbo"
	},
	"dependencies": {
		"@mina-js/providers": "0.1.1",
		"@noble/ciphers": "1.2.1",
		"@noble/curves": "1.8.1",
		"@noble/hashes": "1.7.1",
		"@scure/base": "1.2.4",
		"@scure/bip32": "1.6.2",
		"@scure/bip39": "1.5.4",
		"mina-signer": "3.0.7",
		"zod": "3.24.2"
	},
	"peerDependencies": {
		"typescript": "^5.0.0"
	}
}
