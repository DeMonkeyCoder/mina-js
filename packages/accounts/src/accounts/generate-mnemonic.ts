import { generateMnemonic as generateMnemonic_ } from "@scure/bip39";

type GenerateMnemonicOptions = {
	wordlist: string[];
	strength?: number | undefined;
};

/**
 * @description Generates a random mnemonic phrase with a given wordlist.
 *
 * @param wordlist The wordlist to use for generating the mnemonic phrase.
 * @param strength mnemonic strength 128-256 bits
 *
 * @returns A randomly generated mnemonic phrase.
 */
export function generateMnemonic({
	wordlist,
	strength,
}: GenerateMnemonicOptions): string {
	return generateMnemonic_(wordlist, strength);
}
