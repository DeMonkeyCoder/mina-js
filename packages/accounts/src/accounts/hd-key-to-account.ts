import { sha256 } from "@noble/hashes/sha256";
import { bytesToHex } from "@noble/hashes/utils";
import { base58check } from "@scure/base";
import {
	type HDAccount,
	type HDKey,
	type HDOptions,
	MinaKeyConst,
} from "../types";
import {
	type PrivateKeyToAccountParams,
	privateKeyToAccount,
} from "./private-key-to-account";

export type HDKeyToAccountOptions = HDOptions &
	Partial<PrivateKeyToAccountParams>;

/**
 * @description Creates an Account from a HD Key.
 *
 * @returns A HD Account.
 */
export function hdKeyToAccount(
	hdKey_: HDKey,
	{
		accountIndex = 0,
		addressIndex = 0,
		changeIndex = 0,
		path,
		...options
	}: HDKeyToAccountOptions = {},
): HDAccount {
	const childNode = hdKey_.derive(
		path ||
			`m/${MinaKeyConst.PURPOSE}'/${MinaKeyConst.MINA_COIN_TYPE}'/${accountIndex}'/${changeIndex}/${addressIndex}`,
	);
	if (!childNode.privateKey) throw new Error("Private key not found");
	const childPrivateKeyArray = new Uint8Array(childNode.privateKey);
	childPrivateKeyArray[0] &= 0x3f;
	const childPrivateKey = childPrivateKeyArray.reverse();
	const privateKeyHex = `5a01${bytesToHex(childPrivateKey)}`;
	if (!privateKeyHex) throw new Error("Private key is empty");
	if (!privateKeyHex.match(/.{1,2}/g))
		throw new Error("Failed to split privateKeyHex into bytes");
	const privateKeyBytes = new Uint8Array(
		privateKeyHex.match(/.{1,2}/g)?.map((byte) => Number.parseInt(byte, 16)) ||
			[],
	);
	const account = privateKeyToAccount({
		...options,
		privateKey: base58check(sha256).encode(privateKeyBytes),
	});
	return {
		...account,
		getHdKey: () => childNode,
		source: "hd",
	};
}
