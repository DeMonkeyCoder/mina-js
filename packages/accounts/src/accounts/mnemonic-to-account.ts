import { HDKey } from "@scure/bip32";
import { mnemonicToSeedSync } from "@scure/bip39";

import type { HDAccount, HDOptions } from "../types";
import {
	type HDKeyToAccountOptions,
	hdKeyToAccount,
} from "./hd-key-to-account.js";

export type MnemonicToAccountOptions = HDKeyToAccountOptions;

/**
 * @description Creates an Account from a mnemonic phrase.
 *
 * @returns A HD Account.
 */
export function mnemonicToAccount(
	mnemonic: string,
	opts: HDOptions = {},
): HDAccount {
	const seed = mnemonicToSeedSync(mnemonic);
	return hdKeyToAccount(HDKey.fromMasterSeed(seed), opts);
}
