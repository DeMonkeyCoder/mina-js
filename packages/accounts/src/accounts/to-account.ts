import { type PublicKey, PublicKeySchema } from "@mina-js/utils";
import type {
	AccountSource,
	CustomSource,
	JsonRpcAccount,
	LocalAccount,
} from "../types";

type GetAccountReturnType<accountSource extends AccountSource> =
	| (accountSource extends PublicKey ? JsonRpcAccount : never)
	| (accountSource extends CustomSource ? LocalAccount : never);

/**
 * @description Creates an Account from a custom signing implementation.
 * @returns A Local Account.
 */
export function toAccount<accountSource extends AccountSource>(
	source: accountSource,
): GetAccountReturnType<accountSource> {
	if (typeof source === "string") {
		const publicKey = PublicKeySchema.parse(source);
		return {
			publicKey,
			type: "json-rpc",
		} as GetAccountReturnType<accountSource>;
	}
	const publicKey = PublicKeySchema.parse(source.publicKey);
	return {
		publicKey,
		signMessage: source.signMessage,
		signTransaction: source.signTransaction,
		createNullifier: source.createNullifier,
		signFields: source.signFields,
		source: "custom",
		type: "local",
	} as GetAccountReturnType<accountSource>;
}
