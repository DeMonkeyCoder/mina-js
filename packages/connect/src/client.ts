import "./global-polyfill";
import type {
	Account,
	CreateNullifier,
	SignFields,
	SignMessage,
	SignTransaction,
} from "@mina-js/accounts";
import { createClient } from "@mina-js/klesia-sdk";
import type { PartiallyFormedTransactionProperties } from "@mina-js/shared";
import { match } from "ts-pattern";
import { createStore } from "./store";

export type WalletProviderSlug = "pallad";

export type ProviderSource = "window.mina" | "klesia" | WalletProviderSlug;

export type CreateWalletClientProps = {
	account?: Account;
	network: "mainnet" | "devnet";
	providerSource?: ProviderSource;
};

const getWalletProvider = (providerSource: ProviderSource) => {
	if (providerSource === "window.mina") {
		if (!window.mina) throw new Error("window.mina is not available");
		return window.mina;
	}
	const providerStore = createStore();
	const injectedProvider = providerStore.findProvider({ slug: providerSource });
	if (!injectedProvider)
		throw new Error(`Provider ${providerSource} not found`);
	return injectedProvider.provider;
};

export const createWalletClient = ({
	account,
	network,
	providerSource = "window.mina",
}: CreateWalletClientProps) => {
	const klesiaClient = createClient({ network });
	const getAccounts = async () => {
		return match(providerSource)
			.with("klesia", () => (account ? [account.publicKey] : []))
			.otherwise(async () => {
				const provider = getWalletProvider(providerSource);
				const { result } = await provider.request<"mina_accounts">({
					method: "mina_accounts",
				});
				return result;
			});
	};
	const getBalance = async () => {
		const getBalanceFromInjectedWallet = async () => {
			const provider = getWalletProvider(providerSource);
			const { result } = await provider.request<"mina_getBalance">({
				method: "mina_getBalance",
			});
			return result;
		};
		const getBalanceFromKlesia = async (account: Account) => {
			const { result } = await klesiaClient.request<"mina_getBalance">({
				method: "mina_getBalance",
				params: [account.publicKey],
			});
			return BigInt(result);
		};
		return match(providerSource)
			.with("klesia", async () => {
				if (account) return getBalanceFromKlesia(account);
				return getBalanceFromInjectedWallet();
			})
			.otherwise(getBalanceFromInjectedWallet);
	};
	const getTransactionCount = async () => {
		if (!account)
			throw new Error("Account is required to get transaction count");
		const { result } = await klesiaClient.request<"mina_getTransactionCount">({
			method: "mina_getTransactionCount",
			params: [account.publicKey],
		});
		return BigInt(result);
	};
	const getChainId = async () => {
		return match(providerSource)
			.with("klesia", async () => {
				const { result } = await klesiaClient.request<"mina_chainId">({
					method: "mina_chainId",
				});
				return result;
			})
			.otherwise(async () => {
				const provider = getWalletProvider(providerSource);
				const { result } = await provider.request<"mina_chainId">({
					method: "mina_chainId",
				});
				return result;
			});
	};
	const signTransaction: SignTransaction = async (params) => {
		if (!account) throw new Error("Account is required to sign transaction");
		if (account.type !== "local") throw new Error("Account type not supported");
		return account.signTransaction(params);
	};
	const signMessage: SignMessage = async (params) => {
		if (!account) throw new Error("Account is required to sign message");
		if (account.type !== "local") throw new Error("Account type not supported");
		return account.signMessage(params);
	};
	const signFields: SignFields = async (params) => {
		if (!account) throw new Error("Account is required to sign fields");
		if (account.type !== "local") throw new Error("Account type not supported");
		return account.signFields(params);
	};
	const createNullifier: CreateNullifier = async (params) => {
		if (!account) throw new Error("Account is required to create nullifier");
		if (account.type !== "local") throw new Error("Account type not supported");
		return account.createNullifier(params);
	};
	const estimateFees = async () => {
		const { result } = await klesiaClient.request<"mina_estimateFees">({
			method: "mina_estimateFees",
		});
		return {
			low: BigInt(result.low),
			medium: BigInt(result.medium),
			high: BigInt(result.high),
		};
	};
	const prepareTransactionRequest = async (
		transaction: PartiallyFormedTransactionProperties,
	) => {
		let fee = transaction.fee;
		let nonce = transaction.nonce;
		if (!nonce) {
			nonce = await getTransactionCount();
		}
		if (!fee) {
			fee = BigInt((await estimateFees()).medium);
		}
		return {
			...transaction,
			fee,
			nonce,
		};
	};
	return {
		getAccounts,
		getBalance,
		getTransactionCount,
		getChainId,
		signTransaction,
		signMessage,
		signFields,
		createNullifier,
		estimateFees,
		prepareTransactionRequest,
	};
};
