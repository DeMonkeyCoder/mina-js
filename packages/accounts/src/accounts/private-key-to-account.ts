import {
	NullifierSchema,
	SignedFieldsSchema,
	SignedMessageSchema,
	SignedTransactionSchema,
} from "@mina-js/utils";
import MinaSigner from "mina-signer";
import type { PrivateKeyAccount } from "../types";
import { toAccount } from "./to-account";

export type PrivateKeyToAccountParams = {
	privateKey: string;
	network?: "mainnet" | "testnet";
};

/**
 * @description Creates an Account from a private key.
 *
 * @returns A Private Key Account.
 */
export function privateKeyToAccount({
	privateKey,
	network = "mainnet",
}: PrivateKeyToAccountParams): PrivateKeyAccount {
	const client = new MinaSigner({ network });
	const publicKey = client.derivePublicKey(privateKey);
	const account = toAccount({
		publicKey,
		async signMessage({ message }) {
			return SignedMessageSchema.parse(client.signMessage(message, privateKey));
		},
		async signTransaction({ transaction }) {
			return SignedTransactionSchema.parse(
				client.signTransaction(transaction, privateKey),
			);
		},
		async createNullifier({ message }) {
			return NullifierSchema.parse(
				client.createNullifier(
					message.map((el) => BigInt(el)),
					privateKey,
				),
			);
		},
		async signFields({ fields }) {
			return SignedFieldsSchema.parse(
				client.signFields(
					fields.map((el) => BigInt(el)),
					privateKey,
				),
			);
		},
	});

	return {
		...account,
		source: "privateKey",
	} as PrivateKeyAccount;
}
