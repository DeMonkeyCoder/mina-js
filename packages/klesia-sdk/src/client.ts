import {
	KlesiaNetwork,
	type KlesiaRpc,
	type RpcErrorType,
	type RpcResponseType,
} from "@mina-js/klesia";
import { hc } from "hono/client";
import { match } from "ts-pattern";

type CreateClientProps = {
	network: "mainnet" | "devnet" | "zeko_devnet";
	customUrl?: string;
	throwable?: boolean;
};

const throwRpcError = ({
	code,
	message,
}: { code: number; message: string }) => {
	throw new Error(`${code} - ${message}`);
};

export const createClient = ({
	network,
	customUrl,
	throwable = true,
}: CreateClientProps) => {
	const baseUrl = customUrl
		? customUrl
		: match(KlesiaNetwork.parse(network))
				.with("devnet", () => "https://devnet.klesia.palladians.xyz")
				.with("mainnet", () => "https://mainnet.klesia.palladians.xyz")
				.with("zeko_devnet", () => "https://zeko-devnet.klesia.palladians.xyz")
				.exhaustive();
	const baseClient = hc<KlesiaRpc>(baseUrl, {
		headers: {
			"Access-Control-Allow-Origin": "*",
		},
	});
	const rpcHandler = baseClient.api.$post;
	type RpcRequest = Parameters<typeof rpcHandler>[0];
	const request = async <T extends string>(req: RpcRequest["json"]) => {
		const json = (await (
			await baseClient.api.$post({ json: req })
		).json()) as Extract<RpcResponseType, { method: T }> & {
			error?: RpcErrorType;
		};
		if (!throwable) {
			return json;
		}
		if (json?.error) {
			const { code, message } = json.error;
			return throwRpcError({ code, message });
		}
		return json;
	};
	return {
		request,
	};
};
