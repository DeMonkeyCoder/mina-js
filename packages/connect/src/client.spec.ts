import { describe, expect, it } from "bun:test";
import { privateKeyToAccount } from "@mina-js/accounts";
import { toAccount } from "@mina-js/accounts";
import { Test } from "@mina-js/shared";
import { createWalletClient } from "./client";

const PUBLIC_KEY = "B62qmWKtvNQTtUqo1LxfEEDLyWMg59cp6U7c4uDC7aqgaCEijSc3Hx5";

it("matches snapshot of local source", async () => {
	const account = privateKeyToAccount({
		privateKey: Test.accounts[0].privateKey,
	});
	const client = createWalletClient({ account, network: "devnet" });
	expect(client).toMatchSnapshot();
});

it("matches snapshot of json-rpc source", async () => {
	const account = toAccount(PUBLIC_KEY);
	const client = createWalletClient({ account, network: "devnet" });
	expect(client).toMatchSnapshot();
});

describe("json-rpc source", () => {
	it("returns accounts", async () => {
		const account = toAccount(PUBLIC_KEY);
		const client = createWalletClient({
			account,
			network: "devnet",
			providerSource: "klesia",
		});
		const accounts = await client.getAccounts();
		expect(accounts[0]).toEqual(PUBLIC_KEY);
	});

	// it('returns balance', async () => {
	//   const account = toAccount(PUBLIC_KEY)
	//   const client = createWalletClient({ account, network: 'devnet', providerSource: 'klesia' })
	//   const balance = await client.getBalance()
	//   console.log(balance)
	//   expect(balance).toBeGreaterThanOrEqual(0)
	// })
});
