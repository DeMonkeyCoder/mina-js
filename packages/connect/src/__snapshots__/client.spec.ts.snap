// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot of local source 1`] = `
{
  "createNullifier": [Function: AsyncFunction],
  "getAccounts": [Function: AsyncFunction],
  "getBalance": [Function: AsyncFunction],
  "getNetworkId": [Function: AsyncFunction],
  "getTransactionCount": [Function: AsyncFunction],
  "prepareTransactionRequest": [Function: AsyncFunction],
  "signFields": [Function: AsyncFunction],
  "signMessage": [Function: AsyncFunction],
  "signTransaction": [Function: AsyncFunction],
}
`;

exports[`matches snapshot of json-rpc source 1`] = `
{
  "createNullifier": [Function: AsyncFunction],
  "getAccounts": [Function: AsyncFunction],
  "getBalance": [Function: AsyncFunction],
  "getNetworkId": [Function: AsyncFunction],
  "getTransactionCount": [Function: AsyncFunction],
  "prepareTransactionRequest": [Function: AsyncFunction],
  "signFields": [Function: AsyncFunction],
  "signMessage": [Function: AsyncFunction],
  "signTransaction": [Function: AsyncFunction],
}
`;

exports[`matches snapshot of local source 2`] = `
{
  "createNullifier": [Function: AsyncFunction],
  "getAccounts": [Function: AsyncFunction],
  "getBalance": [Function: AsyncFunction],
  "getNetworkId": [Function: AsyncFunction],
  "getTransactionCount": [Function: AsyncFunction],
  "prepareTransactionRequest": [Function: AsyncFunction],
  "signFields": [Function: AsyncFunction],
  "signMessage": [Function: AsyncFunction],
  "signTransaction": [Function: AsyncFunction],
}
`;

exports[`matches snapshot of json-rpc source 2`] = `
{
  "createNullifier": [Function: AsyncFunction],
  "getAccounts": [Function: AsyncFunction],
  "getBalance": [Function: AsyncFunction],
  "getNetworkId": [Function: AsyncFunction],
  "getTransactionCount": [Function: AsyncFunction],
  "prepareTransactionRequest": [Function: AsyncFunction],
  "signFields": [Function: AsyncFunction],
  "signMessage": [Function: AsyncFunction],
  "signTransaction": [Function: AsyncFunction],
}
`;
